zkevm-circuits/src/evm_circuit/op_execution/pop.rs

```diff
@@ -0,0 +1,305 @@
+                case_selector: success.selector.clone(),
+                // no more for pop
+            },
+            stack_underflow: stack_underflow.selector.clone(),
+            out_of_gas: (
+                out_of_gas.selector.clone(),
+                out_of_gas.resumption.unwrap().gas_available.clone(),
+            ),
+        }
+    }
+
+    fn constraints(
+        &self,
+        op_execution_state_curr: &OpExecutionState<F>,
+        op_execution_state_next: &OpExecutionState<F>,
+    ) -> Vec<Constraint<F>> {
+        let pop = Expression::Constant(F::from_u64(80));
+        let OpExecutionState { opcode, .. } = &op_execution_state_curr;
+        let common_polys = vec![(opcode.expr() - pop.clone())];
+        
+        let success = {
+            // interpreter state transition constraints
+            let one = Expression::Constant(F::from_u64(1));
+            let op_execution_state_transition_constraints = vec![
+                op_execution_state_next.global_counter.expr()
+                    - (op_execution_state_curr.global_counter.expr()
+                        + one.clone()),
+                op_execution_state_next.stack_pointer.expr()
+                    - (op_execution_state_curr.stack_pointer.expr()
+                        + one.clone()),
+                op_execution_state_next.program_counter.expr()
+                    - (op_execution_state_curr.program_counter.expr()
+                        + one.clone()),
+                op_execution_state_next.gas_counter.expr()
+                    - (op_execution_state_curr.gas_counter.expr()
+                        + Expression::Constant(F::from_u64(2))),
+            ];
+
+            let PopSuccessAllocation {
+                case_selector,
+            } = &self.success;
+
+            // TODO: more about pop constraint ?
+
+            Constraint {
+                name: "PopGadget success",
+                selector: case_selector.expr(),
+                polys: [
+                    common_polys.clone(),
+                    op_execution_state_transition_constraints,
+                ]
+                .concat(),
+                lookups: vec![]
+            }
+        };
+
+        let stack_underflow = {
+            let (zero, minus_one) = (
+                Expression::Constant(F::from_u64(1024)),
+                Expression::Constant(F::from_u64(1023)),
+            );
+            let stack_pointer = op_execution_state_curr.stack_pointer.expr();
+            Constraint {
+                name: "PopGadget stack underflow",
+                selector: self.stack_underflow.expr(),
+                polys: vec![
+                    common_polys.clone(),
+                     vec![
+                        (stack_pointer.clone() - zero)
+                            * (stack_pointer - minus_one),
+                    ],
+                ]
+                .concat(),
+                lookups: vec![],
+            }
+        };
+
+        let out_of_gas = {
```

Comment:
I think this does the out-of-gas check as if POP costs 3 gas instead of 2 gas?

Reply:
it is 2, you can refer to gas table in https://github.com/djrtwo/evm-opcode-gas-costs/blob/master/opcode-gas-costs_EIP-150_revision-1e18248_2017-04-12.csv

Reply:
Sorry for being so brief, what I meant was that this seems to still add 3 to the gas counter instead of 2 to check against out-of-gas, so I think this could give a out-of-gas error 1 gas too soon. I believe the code should be like this:
```
let out_of_gas = {
    let (one, two) = (
        Expression::Constant(F::from_u64(1)),
        Expression::Constant(F::from_u64(2)),
    );
    let (case_selector, gas_available) = &self.out_of_gas;
    let gas_overdemand = state_curr.gas_counter.expr() + two.clone()
        - gas_available.expr();
    Constraint {
        name: "Pop out of gas",
        selector: case_selector.expr(),
        polys: [
            common_polys,
            vec![
                (gas_overdemand.clone() - one)
                    * (gas_overdemand.clone() - two),
            ],
        ]
        .concat(),
        lookups: vec![],
    }
};
```